README.md
# cron-comparison

Automated benchmark of all JavaScript `cron` implementations, using [Renovate](https://github.com/renovatebot/renovate) and [GitHub Actions](https://docs.github.com/en/actions).

## Running locally

Run benchmark and show output in console

`npm run benchmark`

Run benchmark and update readme with results

`npm run update-readme`

## Credits

- Hexagon
- sheerlox

## Latest results

> cron-comparison@1.3.1 benchmark
> node --no-warnings src/benchmark.js

Tests performed at 2023-10-09T19:35:28.895Z

Tested libraries ([npm trends](https://npmtrends.com/cron-vs-croner-vs-cronosjs-vs-node-cron-vs-node-schedule)):

- [cron](https://github.com/kelektiv/node-cron) `v3.1.0`
- [croner](https://github.com/hexagon/croner) `v7.0.2`
- [cronosjs](https://github.com/jaclarke/cronosjs) `v1.7.1`
- [node-cron](https://github.com/node-cron/node-cron) `v3.0.2`
- [node-schedule](https://github.com/node-schedule/node-schedule) `v2.1.1`

#### Pattern `0 0 0 L 2 *`
```
Tests

cron            - FAIL  - Error: Unknown alias: l
croner          - OK    - 2024-02-29 00:00:00
cronosjs        - OK    - 2024-02-29 00:00:00
node-cron       - FAIL  - Error: L is a invalid expression for day of month
node-schedule   - OK    - 2024-02-29 00:00:00

Benchmark (only OK)

croner          x 150,794 ops/sec ±8.96% (85 runs sampled)
cronosjs        x 66,798 ops/sec ±15.04% (87 runs sampled)
node-schedule   x 380 ops/sec ±9.81% (83 runs sampled)

Fastest is croner         
```

#### Pattern `0 0 0 * 2 5#L`
```
Tests

cron            - FAIL  - Error: Unknown alias: l
croner          - OK    - 2024-02-23 00:00:00
cronosjs        - FAIL  - Error: Field item invalid format
node-cron       - FAIL  - 1970-01-01 00:00:00
node-schedule   - FAIL  - TypeError: Cannot read properties of null (reading 'nextInvocation')

Benchmark (only OK)

croner          x 59,777 ops/sec ±3.36% (88 runs sampled)

Fastest is croner         
```

#### Pattern `0 0 0 * 12 5#2`
```
Tests

cron            - FAIL  - Error: Field (dayOfWeek) cannot be parsed
croner          - OK    - 2023-12-08 00:00:00
cronosjs        - OK    - 2023-12-08 00:00:00
node-cron       - FAIL  - 1970-01-01 00:00:00
node-schedule   - OK    - 2023-12-08 00:00:00

Benchmark (only OK)

croner          x 133,599 ops/sec ±5.63% (87 runs sampled)
cronosjs        x 58,524 ops/sec ±13.08% (87 runs sampled)
node-schedule   x 3,727 ops/sec ±4.23% (86 runs sampled)

Fastest is croner         
```

#### Pattern `0 0 0 L 12 5-6#3`
```
Tests

cron            - FAIL  - Error: Unknown alias: l
croner          - OK    - 2023-12-15 00:00:00
cronosjs        - OK    - 2023-12-15 00:00:00
node-cron       - FAIL  - Error: L is a invalid expression for day of month
node-schedule   - FAIL  - TypeError: Cannot read properties of null (reading 'nextInvocation')

Benchmark (only OK)

croner          x 86,291 ops/sec ±4.50% (89 runs sampled)
cronosjs        x 56,134 ops/sec ±4.17% (88 runs sampled)

Fastest is croner         
```

#### Pattern `1 2 3 4 5 6`
```
Tests

cron            - OK    - 2024-05-04 03:02:01
croner          - OK    - 2024-05-04 03:02:01
cronosjs        - OK    - 2024-05-04 03:02:01
node-cron       - FAIL  - 1970-01-01 00:00:00
node-schedule   - OK    - 2024-05-04 03:02:01

Benchmark (only OK)

cron            x 6,313 ops/sec ±7.07% (84 runs sampled)
croner          x 160,651 ops/sec ±2.95% (87 runs sampled)
cronosjs        x 55,693 ops/sec ±4.69% (87 runs sampled)
node-schedule   x 2,726 ops/sec ±4.22% (85 runs sampled)

Fastest is croner         
```

#### Pattern `*/3 */3 */3 * * *`
```
Tests

cron            - OK    - 2023-10-09 21:00:00
croner          - OK    - 2023-10-09 21:00:00
cronosjs        - OK    - 2023-10-09 21:00:00
node-cron       - FAIL  - 1970-01-01 00:00:00
node-schedule   - OK    - 2023-10-09 21:00:00

Benchmark (only OK)

cron            x 21,765 ops/sec ±5.38% (83 runs sampled)
croner          x 161,614 ops/sec ±3.57% (86 runs sampled)
cronosjs        x 32,285 ops/sec ±8.05% (88 runs sampled)
node-schedule   x 16,514 ops/sec ±4.31% (88 runs sampled)

Fastest is croner         
```

#### Pattern `0 0 0 29 2 1`
```
Tests

cron            - OK    - 2024-02-05 00:00:00
croner          - OK    - 2024-02-05 00:00:00
cronosjs        - OK    - 2024-02-05 00:00:00
node-cron       - FAIL  - 1970-01-01 00:00:00
node-schedule   - OK    - 2024-02-05 00:00:00

Benchmark (only OK)

cron            x 9,744 ops/sec ±6.65% (84 runs sampled)
croner          x 161,572 ops/sec ±4.47% (88 runs sampled)
cronosjs        x 51,609 ops/sec ±5.92% (81 runs sampled)
node-schedule   x 5,341 ops/sec ±6.21% (82 runs sampled)

Fastest is croner         
```

#### Pattern `0 0 0 29 2 *`
```
Tests

cron            - OK    - 2024-02-29 00:00:00
croner          - OK    - 2024-02-29 00:00:00
cronosjs        - OK    - 2024-02-29 00:00:00
node-cron       - FAIL  - 1970-01-01 00:00:00
node-schedule   - OK    - 2024-02-29 00:00:00

Benchmark (only OK)

cron            x 3,104 ops/sec ±3.65% (86 runs sampled)
croner          x 176,714 ops/sec ±4.69% (86 runs sampled)
cronosjs        x 67,920 ops/sec ±3.43% (89 runs sampled)
node-schedule   x 737 ops/sec ±3.69% (85 runs sampled)

Fastest is croner         
```

#### Pattern `15 15 */3 * * *`
```
Tests

cron            - OK    - 2023-10-09 21:15:15
croner          - OK    - 2023-10-09 21:15:15
cronosjs        - OK    - 2023-10-09 21:15:15
node-cron       - FAIL  - 1970-01-01 00:00:00
node-schedule   - OK    - 2023-10-09 21:15:15

Benchmark (only OK)

cron            x 5,232 ops/sec ±3.66% (86 runs sampled)
croner          x 189,412 ops/sec ±4.44% (88 runs sampled)
cronosjs        x 40,316 ops/sec ±3.56% (89 runs sampled)
node-schedule   x 2,313 ops/sec ±4.69% (84 runs sampled)

Fastest is croner         
```

#### Pattern `15 15 */3 */10 10 *`
```
Tests

cron            - OK    - 2023-10-11 00:15:15
croner          - OK    - 2023-10-11 00:15:15
cronosjs        - OK    - 2023-10-11 00:15:15
node-cron       - FAIL  - 1970-01-01 00:00:00
node-schedule   - OK    - 2023-10-11 00:15:15

Benchmark (only OK)

cron            x 5,129 ops/sec ±4.01% (86 runs sampled)
croner          x 158,104 ops/sec ±6.95% (81 runs sampled)
cronosjs        x 55,083 ops/sec ±3.85% (87 runs sampled)
node-schedule   x 2,622 ops/sec ±5.25% (80 runs sampled)

Fastest is croner         
```

#### Pattern `15 15 */3 * 10 SUN,MON,TUE`
```
Tests

cron            - OK    - 2023-10-09 21:15:15
croner          - OK    - 2023-10-09 21:15:15
cronosjs        - OK    - 2023-10-09 21:15:15
node-cron       - FAIL  - 1970-01-01 00:00:00
node-schedule   - OK    - 2023-10-09 21:15:15

Benchmark (only OK)

cron            x 5,131 ops/sec ±4.00% (88 runs sampled)
croner          x 111,913 ops/sec ±3.74% (88 runs sampled)
cronosjs        x 42,010 ops/sec ±3.53% (88 runs sampled)
node-schedule   x 2,275 ops/sec ±6.14% (83 runs sampled)

Fastest is croner         
```

## Test summary

| Library | OK | FAIL | % OK |
| :---: | :--: | :---: | :-----: |
| cron | 7 | 4 | 63.63636363636363% |
| croner | 11 | 0 | 100% |
| cronosjs | 10 | 1 | 90.9090909090909% |
| node-cron | 0 | 11 | 0% |
| node-schedule | 9 | 2 | 81.81818181818183% |



CHANGELOG.md
## [1.3.1](https://github.com/sheerlox/cron-comparison/compare/v1.3.0...v1.3.1) (2023-10-09)

### 🐛 Bug Fixes

- **libs:** update dependency cron to v3.1.0 ([#15](https://github.com/sheerlox/cron-comparison/issues/15)) ([85d5db1](https://github.com/sheerlox/cron-comparison/commit/85d5db18b9d39d11f864191a2836c5fd55bc8921))

### ⚙️ Continuous Integrations

- **renovate:** stop grouping for improved changelog ([442049b](https://github.com/sheerlox/cron-comparison/commit/442049b14d5ffe64a39681c2a04df7400d91d9d7))

## [1.3.0](https://github.com/sheerlox/cron-comparison/compare/v1.2.3...v1.3.0) (2023-10-02)

### ✨ Features

- add npm trends link ([7e22f93](https://github.com/sheerlox/cron-comparison/commit/7e22f93b09c101b810beff3d1f59a2035d322638))

### ♻️ Chores

- **deps:** update semantic-release related packages ([#13](https://github.com/sheerlox/cron-comparison/issues/13)) ([6a2a0bd](https://github.com/sheerlox/cron-comparison/commit/6a2a0bddd7dd5bddf808ed5e38fd8b818fb2073e))

## [1.2.3](https://github.com/sheerlox/cron-comparison/compare/v1.2.2...v1.2.3) (2023-09-30)

### 📚 Documentation

- **readme:** update template ([683c483](https://github.com/sheerlox/cron-comparison/commit/683c483ac6d056aa01e474049eb4f1d782e53745))

## [1.2.2](https://github.com/sheerlox/cron-comparison/compare/v1.2.1...v1.2.2) (2023-09-30)

### 🐛 Bug Fixes

- **libs:** update benchmarked libs ([#11](https://github.com/sheerlox/cron-comparison/issues/11)) ([bde6ac1](https://github.com/sheerlox/cron-comparison/commit/bde6ac187a57e64ecd42740d3ce200efa6b2ff85))

## [1.2.1](https://github.com/sheerlox/cron-comparison/compare/v1.2.0...v1.2.1) (2023-09-30)

### 📚 Documentation

- **readme:** update template ([7fe4770](https://github.com/sheerlox/cron-comparison/commit/7fe4770bd4b0c7f9043bcd3f5ec444be63df6b03))

### ⚙️ Continuous Integrations

- **action:** use CI token to bypass branch protection ([7449003](https://github.com/sheerlox/cron-comparison/commit/7449003f034ac252b1624f4b9cffe7a1af194b9c))

## [1.2.0](https://github.com/sheerlox/cron-comparison/compare/v1.1.1...v1.2.0) (2023-09-29)

### ✨ Features

- add librairies version numbers to results ([f2bd722](https://github.com/sheerlox/cron-comparison/commit/f2bd722c18dfafb34fa2c7ded943cec1831ae391))
- sort librairies alphabetically ([a902978](https://github.com/sheerlox/cron-comparison/commit/a9029785fd935e69215ff80339a482f647f81519))

### 🐛 Bug Fixes

- fix best count increment ([fa354f8](https://github.com/sheerlox/cron-comparison/commit/fa354f85b956c0dcd757d3797c8e298cceb24b7a))

### 📦 Code Refactoring

- centralize test subjects loading ([f97c93e](https://github.com/sheerlox/cron-comparison/commit/f97c93e59857eeeff99950e608e4c16b354f1130))
- use ES syntax for every module ([8bda71a](https://github.com/sheerlox/cron-comparison/commit/8bda71a9a5d57e068634f9303e20a9cf4edfb710))

## [1.1.1](https://github.com/sheerlox/cron-comparison/compare/v1.1.0...v1.1.1) (2023-09-29)

### 🐛 Bug Fixes

- **libs:** pin dependencies ([#8](https://github.com/sheerlox/cron-comparison/issues/8)) [skip ci] ([edf04cb](https://github.com/sheerlox/cron-comparison/commit/edf04cb2a571c5f58c67d4c7af1e93ebdd28423a))
- **libs:** update benchmarked libs ([#10](https://github.com/sheerlox/cron-comparison/issues/10)) ([33f7450](https://github.com/sheerlox/cron-comparison/commit/33f745040ce91f191527a8f9be939e00acab2f22))

### ⚙️ Continuous Integrations

- **action:** run tests on PRs to main branch ([abd8e29](https://github.com/sheerlox/cron-comparison/commit/abd8e2992df57674aeea3c456a005a017cd88467))
- **action:** update actions/checkout action to v4 ([#5](https://github.com/sheerlox/cron-comparison/issues/5)) ([a9dae67](https://github.com/sheerlox/cron-comparison/commit/a9dae676487541268f249402007dba24fa3858de))
- **renovate:** update benchmarked libraries group config ([6bf099e](https://github.com/sheerlox/cron-comparison/commit/6bf099e0a285f9a00bb19cc06da462acdf911b60))

### ♻️ Chores

- add libs label to renovate config ([e13852f](https://github.com/sheerlox/cron-comparison/commit/e13852fe63e5d34f65267adb87eded8df49475e3))
- configure renovate ([3a435b6](https://github.com/sheerlox/cron-comparison/commit/3a435b6864f36e66e9ea8e674d1029ea57b1a3c9))
- **deps:** pin dependencies ([#2](https://github.com/sheerlox/cron-comparison/issues/2)) ([49129e0](https://github.com/sheerlox/cron-comparison/commit/49129e01a18eb5fddcf6a0672094719b3b661123))
- **deps:** update [@semantic-release](https://github.com/semantic-release) packages ([8b20f75](https://github.com/sheerlox/cron-comparison/commit/8b20f75768b29f968b6fa26c96d2a91005a72174))

## [1.1.0](https://github.com/sheerlox/cron-comparison/compare/v1.0.2...v1.1.0) (2023-09-29)

### ✨ Features

- make date formatting locale agnostic ([2f702a5](https://github.com/sheerlox/cron-comparison/commit/2f702a5797058c9d5682f92a6492df651adae77b))
- remove unused method & add test file to check API changes ([cf5de8d](https://github.com/sheerlox/cron-comparison/commit/cf5de8d1dccb6cd84c8fdbabdf7a9c8447b654ab))

### ⚙️ Continuous Integrations

- **action:** add release workflow ([34fec8d](https://github.com/sheerlox/cron-comparison/commit/34fec8da740e923e3860f0e3529f3bb1cbb56cc4))
- **action:** add test workflow ([6fe5fd0](https://github.com/sheerlox/cron-comparison/commit/6fe5fd0de3d1d7830aee9b8415faf7e46cbbcc61))

### 💎 Styles

- setup prettier ([f02b5cc](https://github.com/sheerlox/cron-comparison/commit/f02b5cc4de17255373d7e1aae5ae8196dc376862))

## [1.0.2](https://github.com/hexagon/cron-comparison/compare/de2f047...57132f7) (2023-10-09)

### Initial work

index.js
// Prepare require for modules who require it...
import { createRequire } from 'node:module';
const require = createRequire(import.meta.url);

// Prepare benchmark
var Benchmark = require('benchmark');

// Commonjs modules 
const NodeSchedule = require("./interfaces/node-schedule.cjs");
const Cron = require("./interfaces/cron.cjs");
const NodeCron = require("./interfaces/nodecron.cjs");

// ES modules
import Croner from "./interfaces/croner.js";
import CronosJS from "./interfaces/cronosjs.js";

console.log("Tests performed at " + new Date().toISOString());
const summary = {};

for (const pattern of ["0 0 0 L 2 *", "1 2 3 4 5 6", "*/3 */3 */3 * * *", "0 0 0 29 2 1", "0 0 0 29 2 *","15 15 */3 * * *","15 15 */3 */10 10 *","15 15 */3 * 10 SUN,MON,TUE"]) {

    console.log("");
    console.log("#### Pattern '" + pattern + "'");
    console.log("```");
    console.log("Tests");
    console.log("");

    let reports = [];

    for (const scheduler of [Croner, NodeCron, NodeSchedule, CronosJS, Cron]) {
        let 
            job = new scheduler(pattern),
            result = { id: job.id, scheduler };
        try {
            job.init();
            result.result = new Date(job.next()).toLocaleString();
            job.stop();
        } catch (e) {
            result.result = undefined;
            result.message = e.toString();
        }
        reports.push(result);
    }

    // Guess correct answer from popular demand
    let correct = {},
        bestCount = 0,
        best;
    for(const report of reports) {
      if (report.result) {
        correct[report.result] = correct[report.result] ? correct[report.result]++ : 1;
        if (correct[report.result] > bestCount) {
          bestCount = correct[report.result];
          best = report.result;
        }
      }
    }

    // Print test results
    for(const report of reports) {
      report.correct = report?.result == best;
      console.log(report.id.padEnd(15, " ") + " - " + (report.correct ? "OK   " : "FAIL ") + " - " + (report.correct ? report.result : (report.result || report.message)));
      summary[report.id] = summary[report.id] || {};
      summary[report.id].ok = (summary[report.id].ok || 0) + (report.correct ?  1 : 0);
      summary[report.id].fail = (summary[report.id].fail || 0) + (!report.correct ? 1 : 0);
    }

    console.log("");
    console.log("Benchmark (only OK)");
    console.log("");

    var bm = new Benchmark.Suite;

    for (const report of reports) {
      if( report.correct ) {
        const job = new report.scheduler(pattern);
        bm.add(job.id.padEnd(15, " "), function() { job.init(pattern); job.next(); job.stop(); });
      } 
    }
    
    bm
      .on('cycle', function(event) {
        console.log(String(event.target));
      })
      .on('complete', function() {
        console.log("");
        console.log("Fastest is " + this.filter('fastest').map('name'));
      })  
      .run({ 'async' : false });

    console.log("```");
}

console.log("");
console.log("## Test summary");
console.log("");
console.log("| Library | OK | FAIL | % OK |");
console.log("| :---: | :--: | :---: | :-----: |");
for(const report in summary) {
  console.log("| " + report + " | " + summary[report].ok + " | " + summary[report].fail + " | " + (summary[report].ok / Math.round((summary[report].ok + summary[report].fail))*100) + "% |" );
}
console.log("");

src/benchmark.js
import Benchmark from "benchmark";

import schedulers from "./interfaces/index.js";

const summary = {};

console.log("Tests performed at " + new Date().toISOString() + "\n");

const npmTrendsLink = `https://npmtrends.com/${schedulers
  .map(({ id }) => id)
  .join("-vs-")}`;

console.log(`Tested libraries ([npm trends](${npmTrendsLink})):\n`);
for (const scheduler of schedulers) {
  console.log(
    `- [${scheduler.id}](${scheduler.url}) \`v${scheduler.version}\``,
  );
}

for (const pattern of [
  "0 0 0 L 2 *",
  "0 0 0 * 2 5#L",
  "0 0 0 * 12 5#2",
  "0 0 0 L 12 5-6#3",
  "1 2 3 4 5 6",
  "*/3 */3 */3 * * *",
  "0 0 0 29 2 1",
  "0 0 0 29 2 *",
  "15 15 */3 * * *",
  "15 15 */3 */10 10 *",
  "15 15 */3 * 10 SUN,MON,TUE",
]) {
  console.log("");
  console.log("#### Pattern `" + pattern + "`");
  console.log("```");
  console.log("Tests");
  console.log("");

  let reports = [];

  for (const scheduler of schedulers) {
    let job = new scheduler.interface(pattern),
      result = { id: scheduler.id, scheduler: scheduler.interface };
    try {
      job.init();
      result.result = formatDateString(new Date(job.next()));
      job.stop();
    } catch (e) {
      result.result = undefined;
      result.message = e.toString();
    }
    reports.push(result);
  }

  // Guess correct answer from popular demand
  let correct = {},
    bestCount = 0,
    best;
  for (const report of reports) {
    if (report.result) {
      correct[report.result] = correct[report.result]
        ? correct[report.result] + 1
        : 1;
      if (correct[report.result] > bestCount) {
        bestCount = correct[report.result];
        best = report.result;
      }
    }
  }

  // Print test results
  for (const report of reports) {
    report.correct = report?.result == best;
    console.log(
      report.id.padEnd(15, " ") +
        " - " +
        (report.correct ? "OK   " : "FAIL ") +
        " - " +
        (report.correct ? report.result : report.result || report.message),
    );
    summary[report.id] = summary[report.id] || {};
    summary[report.id].ok =
      (summary[report.id].ok || 0) + (report.correct ? 1 : 0);
    summary[report.id].fail =
      (summary[report.id].fail || 0) + (!report.correct ? 1 : 0);
  }

  console.log("");
  console.log("Benchmark (only OK)");
  console.log("");

  var bm = new Benchmark.Suite();

  for (const report of reports) {
    if (report.correct) {
      const job = new report.scheduler(pattern);
      bm.add(report.id.padEnd(15, " "), function () {
        job.init(pattern);
        job.next();
        job.stop();
      });
    }
  }

  bm.on("cycle", function (event) {
    console.log(String(event.target));
  })
    .on("complete", function () {
      console.log("");
      console.log("Fastest is " + this.filter("fastest").map("name"));
    })
    .run({ async: false });

  console.log("```");
}

console.log("");
console.log("## Test summary");
console.log("");
console.log("| Library | OK | FAIL | % OK |");
console.log("| :---: | :--: | :---: | :-----: |");
for (const report in summary) {
  console.log(
    "| " +
      report +
      " | " +
      summary[report].ok +
      " | " +
      summary[report].fail +
      " | " +
      (summary[report].ok /
        Math.round(summary[report].ok + summary[report].fail)) *
        100 +
      "% |",
  );
}
console.log("");

function formatDateString(date) {
  const year = date.getFullYear();
  const month = String(date.getMonth() + 1).padStart(2, "0");
  const day = String(date.getDate()).padStart(2, "0");
  const hours = String(date.getHours()).padStart(2, "0");
  const minutes = String(date.getMinutes()).padStart(2, "0");
  const seconds = String(date.getSeconds()).padStart(2, "0");

  return `${year}-${month}-${day} ${hours}:${minutes}:${seconds}`;
}
